File: Numerical_method/Python/lab7/1_3.py
# Simpson's 1/3 Rule

def simpsons_1_3_rule(f, a, b, n):
    # Calculate step size
    h = (b - a) / n

    # Initial sum
    s = f(a) + f(b)

    # Compute the sum of even-indexed terms
    for i in range(1, n, 2):
        s += 4 * f(a + i * h)

    # Compute the sum of odd-indexed terms
    for i in range(2, n, 2):
        s += 2 * f(a + i * h)

    # Final result
    return s * h / 3

# Example function to integrate (you can replace this with your own function)
def f(x):
    return 1 / (1 + x ** 2)

# Input section
lower_limit = float(input("Enter lower limit of integration: "))
upper_limit = float(input("Enter upper limit of integration: "))
sub_intervals = int(input("Enter number of subintervals: "))

# Calculate the integral using Simpson's 1/3 Rule
result = simpsons_1_3_rule(f, lower_limit, upper_limit, sub_intervals)

print(f"Integration result by Simpson's 1/3 method is: {result:.6F}")

File: Numerical_method/C++/lab7/1_3.cpp
// Simpson's 1/3 Rule

#include <iostream>
#include <cmath>

// Define the function to integrate
double f(double x) {
    return 1 / (1 + x * x);
}

// Simpson's 1/3 Rule function
double simpsons_1_3_rule(double a, double b, int n) {
    // Ensure n is even
    if (n % 2 != 0) {
        std::cerr << "Error: Number of subintervals must be even." << std::endl;
        return 0;
    }

    // Calculate step size
    double h = (b - a) / n;

    // Initial sum
    double s = f(a) + f(b);

    // Compute the sum of even-indexed terms
    for (int i = 1; i < n; i += 2) {
        s += 4 * f(a + i * h);
    }

    // Compute the sum of odd-indexed terms
    for (int i = 2; i < n; i += 2) {
        s += 2 * f(a + i * h);
    }

    // Final result
    return s * h / 3;
}

int main() {
    // Input section
    double lower_limit, upper_limit;
    int sub_intervals;

    std::cout << "Enter lower limit of integration: ";
    std::cin >> lower_limit;

    std::cout << "Enter upper limit of integration: ";
    std::cin >> upper_limit;

    std::cout << "Enter number of subintervals (must be even): ";
    std::cin >> sub_intervals;

    // Calculate the integral using Simpson's 1/3 Rule
    double result = simpsons_1_3_rule(lower_limit, upper_limit, sub_intervals);

    // Output result
    std::cout << "Integration result by Simpson's 1/3 method is: " << result << std::endl;

    return 0;
}
